### ISSUE TITLE
`NoReturnInBeginEndBlocks` inconsistency when assigning to an instance variable

### ISSUE THREAD

**Issue by Drowze** [rubocop/rubocop#14706]
## Steps to reproduce the problem

Given file:
```ruby
# reports offense as expected
foo_1 = begin
  return if rand(0..1) == 1

  42
end

# reports offense as expected
foo_2 ||= begin
  return if rand(0..1) == 1

  42
end

# BUG: does NOT report offense
@foo_3 = begin
  return if rand(0..1) == 1

  42
end

# reports offense as expected
@foo_4 ||= begin
  return if rand(0..1) == 1

  42
end
```

Running rubocop:
```bash
bundle exec rubocop --only Lint/NoReturnInBeginEndBlocks test.rb
```

## Expected behavior

I expected to see 4 offenses reported in file example above.

## Actual behavior

Only three offenses were reported. Namely, the 3rd assignment (`@foo = begin`) does not report an offense.

Full output:
```
Inspecting 1 file
W

Offenses:

test.rb:3:3: W: Lint/NoReturnInBeginEndBlocks: Do not return in begin..end blocks in assignment contexts.
  return if rand(0..1) == 1
  ^^^^^^
test.rb:10:3: W: Lint/NoReturnInBeginEndBlocks: Do not return in begin..end blocks in assignment contexts.
  return if rand(0..1) == 1
  ^^^^^^
test.rb:24:3: W: Lint/NoReturnInBeginEndBlocks: Do not return in begin..end blocks in assignment contexts.
  return if rand(0..1) == 1
  ^^^^^^

1 file inspected, 3 offenses detected
```

## RuboCop version

```
$ bundle exec rubocop -V
1.81.7 (using Parser 3.3.10.0, Prism 1.6.0, rubocop-ast 1.48.0, analyzing as Ruby 3.4, running on ruby 3.4.1) [arm64-darwin24]
  - rubocop-rake 0.7.1
  - rubocop-rails 2.34.2
  - rubocop-rspec 3.8.0
  - rubocop-performance 1.26.1
  - rubocop-factory_bot 2.28.0
  - rubocop-overhaul 0.5.0
  - rubocop-rspec_rails 2.32.0
  - rubocop-thread_safety 0.7.3
```
